File: mergeSort.js
const mergeSort = (input) => merge(split(input));

const split = (input) => {
  if (input.length === 1) { 
    return input;
  }

  if (input.length === 2) {
    if (input[0] === undefined || input[0] === null) {
      return [input[1], input[0]];
    } else if (input[1] === undefined || input[1] === null) {
      return [input[0], input[1]];
    } else if (input[1] > input[0]) {
      return input;
    } else {
      return [input[1], input[0]];
    }
  }

  return merge(
    split(input.slice(0, input.length / 2)),
    split(input.slice(input.length / 2)),
  );
};

const merge = (left, right) => {
  if (!right) { return left; }

  let result = [];
  let idxLeft = 0;
  let idxRight = 0;
  let leftVal;
  let rightVal;

  while (true) {
    leftVal = left[idxLeft];
    rightVal = right[idxRight];

    if (idxLeft === left.length) {
      result = result.concat(right.slice(idxRight));
      break;
    } else if (idxRight === right.length) {
      result = result.concat(left.slice(idxLeft));
      break;
    } else if (leftVal === undefined || leftVal === null) {
      result.push(rightVal);
      idxRight += 1;
    } else if (rightVal === undefined || rightVal === null) {
      result.push(leftVal);
      idxLeft += 1;
    } else if (leftVal > rightVal) {
      result.push(rightVal);
      idxRight += 1;
    } else {
      result.push(leftVal);
      idxLeft += 1;
    }
  }

  return result;
};

module.exports.mergeSort = mergeSort;

File: merge_sort.py
from pprint import pprint

def merge_sort(items):
  return merge(split(items), None)

def split(items):
  if len(items) == 1:
    return items

  if len(items) == 2:
    if items[1] > items[0]:
      return items
    else:
      return [items[1], items[0]]

  return merge(
    split(items[0 : int(len(items) / 2)]),
    split(items[int(len(items) / 2) : len(items)]),
  )

def merge(left, right):
  if right == None:
    return left

  result = []
  idx_left = 0
  idx_right = 0

  while True:
    if idx_left == len(left):
      return result + right[idx_right:]
    elif idx_right == len(right):
      return result + left[idx_left:]
    elif left[idx_left] < right[idx_right]:
      result.append(left[idx_left])
      idx_left += 1
    else:
      result.append(right[idx_right])
      idx_right += 1

pprint(merge_sort([1, 2, 3]))
pprint(merge_sort([2, 1, 3]))
pprint(merge_sort(['a', 'b', 'c']))
pprint(merge_sort(['b', 'a', 'c']))
pprint(merge_sort(list('The quick brown fox jumps over the lazy dog')))
